src/main/haskell/matching-logic/src/Kore/MatchingLogic/HilbertProof.hs

```diff
@@ -0,0 +1,65 @@
+module Kore.MatchingLogic.HilbertProof
+  (Proof(index,claims,derivations)
+  ,ProofSystem(..)
+  ,emptyProof
+  ,add
+  ,derive
+  ,renderProof
+  )   where
+import Data.List(lookup,delete,nub)
+import Control.Monad(guard)
+import Data.Map.Strict(Map)
+import qualified Data.Map.Strict as Map
+import Data.Sequence(Seq,(|>))
+import qualified Data.Sequence as Seq
+import Data.Foldable
+
+data Proof ix rule formula =
+  Proof
+  { index :: Map ix (Int,formula)
+  , claims :: Seq (ix,formula)
+  , derivations :: Map ix (rule,[(ix,formula)])
```

Comment:
I left the representation redundant to avoid clients needing to look things up by index. I think the best thing to do here is to stop exposing the field accessors directly, and have lookup up functions that will include formulas in the results when convenient, whatever the internal representation of a `Proof`. (If we want to support adding a new formula in the middle of a proof, we might like some kind of order maintenance data structure).